File: tests/ts/test_request_response_sdk.ts
/**
 * Request / response tests for the TypeScript StructFrameSdk.
 *
 * Uses MockTransport + real frame encoding.  Responses are injected via
 * setTimeout so they arrive after request() has registered its subscription
 * and the returned Promise is awaiting.
 *
 * Test matrix
 * -----------
 * - Basic request/response: request() resolves with the first matching response.
 * - Timeout: request() rejects when no response arrives.
 * - match predicate: responses are filtered by a correlation field.
 * - Concurrent requests: two in-flight requests resolve independently.
 * - Subscription cleanup: handler removed after success and after timeout.
 */

import { BasicTypesMessage, getMessageInfo } from '../generated/ts/serialization-test.structframe';
import {
  BufferWriter,
  ProfileStandardConfig,
} from '../generated/ts/frame-profiles';
import { StructFrameSdk } from '../generated/ts/struct-frame-sdk/struct-frame-sdk';
import { ITransport } from '../generated/ts/struct-frame-sdk/transport';

// =============================================================================
// Mock transport
// =============================================================================

class MockTransport implements ITransport {
  private dataCallback?: (data: Uint8Array) => void;
  private errorCallback?: (error: Error) => void;
  private closeCallback?: () => void;

  public readonly sentData: Uint8Array[] = [];
  private _connected = false;

  async connect(): Promise<void> { this._connected = true; }
  async disconnect(): Promise<void> { this._connected = false; }
  async send(data: Uint8Array): Promise<number> { this.sentData.push(data); return data.length; }

  isConnected(): boolean { return this._connected; }

  onData(callback: (data: Uint8Array) => void): void { this.dataCallback = callback; }
  onError(callback: (error: Error) => void): void { this.errorCallback = callback; }
  onClose(callback: () => void): void { this.closeCallback = callback; }

  injectData(data: Uint8Array): void { this.dataCallback?.(data); }
}

// =============================================================================
// Helpers
// =============================================================================

function encodeMsg(regularInt: number, flag: boolean, mediumInt = 0): Uint8Array {
  const msg = new BasicTypesMessage({ regularInt, flag, mediumInt });
  const writer = new BufferWriter(ProfileStandardConfig, 512);
  writer.write(msg);
  return writer.data();
}

function makeSdk(transport: MockTransport): StructFrameSdk {
  return new StructFrameSdk({ transport, profile: ProfileStandardConfig, getMessageInfo });
}

function injectAfter(transport: MockTransport, data: Uint8Array, delayMs = 30): void {
  setTimeout(() => transport.injectData(data), delayMs);
}

// =============================================================================
// Test infrastructure
// =============================================================================

let passed = 0;
let failed = 0;

function assert(name: string, condition: boolean): void {
  console.log(`  ${condition ? 'PASS' : 'FAIL'}  ${name}`);
  if (condition) passed++; else failed++;
}

// =============================================================================
// Tests
// =============================================================================

async function testBasicRequestResponse(): Promise<void> {
  const transport = new MockTransport();
  const sdk = makeSdk(transport);

  // Register codec so handler receives deserialized objects.
  sdk.registerCodec({
    getMsgId: () => BasicTypesMessage._msgid,
    deserialize: (data) => BasicTypesMessage.deserialize(Buffer.from(data)),
  });

  injectAfter(transport, encodeMsg(10, true));

  const result = await sdk.request<BasicTypesMessage>(
    new BasicTypesMessage({ regularInt: 10 }),
    BasicTypesMessage._msgid,
    { timeout: 2 },
  );

  assert('basic: request() resolves', result !== null && result !== undefined);
  assert('basic: result is BasicTypesMessage', result instanceof BasicTypesMessage);
  assert('basic: flag field correct', (result as any).flag === true);
  assert('basic: request was sent', transport.sentData.length === 1);
}

async function testTimeout(): Promise<void> {
  const transport = new MockTransport();
  const sdk = makeSdk(transport);

  let rejected = false;
  try {
    await sdk.request<BasicTypesMessage>(
      new BasicTypesMessage({ regularInt: 99 }),
      BasicTypesMessage._msgid,
      { timeout: 0.1 },
    );
  } catch (err) {
    rejected = true;
    assert('timeout: error message mentions msgId',
      err instanceof Error && err.message.includes(String(BasicTypesMessage._msgid)));
  }
  assert('timeout: Promise rejects', rejected);
  assert('timeout: request was still sent', transport.sentData.length === 1);
}

async function testMatchPredicateFiltersResponses(): Promise<void> {
  const transport = new MockTransport();
  const sdk = makeSdk(transport);

  sdk.registerCodec({
    getMsgId: () => BasicTypesMessage._msgid,
    deserialize: (data) => BasicTypesMessage.deserialize(Buffer.from(data)),
  });

  // Inject a wrong response first, then the correct one.
  setTimeout(() => {
    transport.injectData(encodeMsg(99, false));   // wrong
    setTimeout(() => transport.injectData(encodeMsg(7, true)), 30);  // correct
  }, 20);

  const result = await sdk.request<BasicTypesMessage>(
    new BasicTypesMessage({ regularInt: 7 }),
    BasicTypesMessage._msgid,
    {
      match: (r) => (r as any).regularInt === 7,
      timeout: 2,
    },
  );

  assert('match: non-matching response ignored', result !== null);
  assert('match: correct response returned', (result as any).regularInt === 7);
  assert('match: flag field from matching response', (result as any).flag === true);
}

async function testConcurrentRequestsWithMatch(): Promise<void> {
  const transport = new MockTransport();
  const sdk = makeSdk(transport);

  sdk.registerCodec({
    getMsgId: () => BasicTypesMessage._msgid,
    deserialize: (data) => BasicTypesMessage.deserialize(Buffer.from(data)),
  });

  // Inject responses in reverse order after both requests are parked.
  setTimeout(() => {
    transport.injectData(encodeMsg(2, false));  // for request B first
    transport.injectData(encodeMsg(1, true));   // for request A
  }, 40);

  const [resultA, resultB] = await Promise.all([
    sdk.request<BasicTypesMessage>(
      new BasicTypesMessage({ regularInt: 1 }),
      BasicTypesMessage._msgid,
      { match: (r) => (r as any).regularInt === 1, timeout: 2 },
    ),
    sdk.request<BasicTypesMessage>(
      new BasicTypesMessage({ regularInt: 2 }),
      BasicTypesMessage._msgid,
      { match: (r) => (r as any).regularInt === 2, timeout: 2 },
    ),
  ]);

  assert('concurrent: request A resolved correctly', (resultA as any).regularInt === 1);
  assert('concurrent: request B resolved correctly', (resultB as any).regularInt === 2);
  assert('concurrent: A flag correct', (resultA as any).flag === true);
  assert('concurrent: B flag correct', (resultB as any).flag === false);
}

async function testSubscriptionRemovedAfterSuccess(): Promise<void> {
  const transport = new MockTransport();
  const sdk = makeSdk(transport);

  sdk.registerCodec({
    getMsgId: () => BasicTypesMessage._msgid,
    deserialize: (data) => BasicTypesMessage.deserialize(Buffer.from(data)),
  });

  injectAfter(transport, encodeMsg(3, false));
  await sdk.request<BasicTypesMessage>(
    new BasicTypesMessage({ regularInt: 3 }),
    BasicTypesMessage._msgid,
    { timeout: 2 },
  );

  const handlers = (sdk as any).messageHandlers.get(BasicTypesMessage._msgid) ?? [];
  assert('cleanup: subscription removed after success', handlers.length === 0);
}

async function testSubscriptionRemovedAfterTimeout(): Promise<void> {
  const transport = new MockTransport();
  const sdk = makeSdk(transport);

  try {
    await sdk.request<BasicTypesMessage>(
      new BasicTypesMessage({ regularInt: 0 }),
      BasicTypesMessage._msgid,
      { timeout: 0.05 },
    );
  } catch {
    // expected
  }

  const handlers = (sdk as any).messageHandlers.get(BasicTypesMessage._msgid) ?? [];
  assert('cleanup: subscription removed after timeout', handlers.length === 0);
}

// =============================================================================
// Main
// =============================================================================

async function main(): Promise<void> {
  console.log();
  console.log('========================================');
  console.log('REQUEST/RESPONSE TESTS - TypeScript');
  console.log('========================================');
  console.log();

  await testBasicRequestResponse();
  await testTimeout();
  await testMatchPredicateFiltersResponses();
  await testConcurrentRequestsWithMatch();
  await testSubscriptionRemovedAfterSuccess();
  await testSubscriptionRemovedAfterTimeout();

  console.log();
  console.log('========================================');
  console.log(`Summary: ${passed}/${passed + failed} tests passed`);
  console.log('========================================');
  console.log();

  process.exit(failed > 0 ? 1 : 0);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: tests/py/test_request_response_sdk.py
#!/usr/bin/env python3
"""
Request / response tests for the Python StructFrameSdk.

Each test drives a single SDK instance against a MockTransport.  A background
thread injects the response frame after a short delay to let request() block
first, then unblocks it.  This mirrors real async I/O without requiring a real
network connection.

Test matrix
-----------
- Basic request/response: request() returns the first matching response.
- Timeout: request() raises TimeoutError when no response arrives.
- match predicate: concurrent requests with different correlation IDs are
  resolved independently.
- request_raw(): returns raw payload bytes when no response class is needed.
- Subscription cleanup: the one-shot handler is removed after a successful
  request and after a timeout.
"""

import sys
import os
import threading
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'generated', 'py'))

_sdk_dir = os.path.join(
    os.path.dirname(__file__), '..', '..', 'src', 'struct_frame', 'boilerplate', 'py'
)
sys.path.insert(0, _sdk_dir)

from struct_frame_sdk.struct_frame_sdk import StructFrameSdk, StructFrameSdkConfig
from struct_frame_sdk.transport import ITransport

from frame_profiles import (
    BufferWriter,
    PROFILE_STANDARD_CONFIG,
)
from struct_frame.generated.serialization_test import (
    BasicTypesMessage,
    SerializationTestMessage,
    get_message_info,
)


# ---------------------------------------------------------------------------
# Mock transport
# ---------------------------------------------------------------------------

class MockTransport(ITransport):
    def __init__(self):
        self._data_cb = None
        self._error_cb = None
        self._close_cb = None
        self._connected = False
        self.sent_data = []

    def connect(self): self._connected = True
    def disconnect(self): self._connected = False
    def send(self, data: bytes): self.sent_data.append(bytes(data)); return len(data)
    def set_data_callback(self, cb): self._data_cb = cb
    def set_error_callback(self, cb): self._error_cb = cb
    def set_close_callback(self, cb): self._close_cb = cb
    def is_connected(self): return self._connected

    def inject_data(self, data: bytes):
        if self._data_cb:
            self._data_cb(data)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def make_sdk(transport):
    return StructFrameSdk(StructFrameSdkConfig(
        transport=transport,
        profile=PROFILE_STANDARD_CONFIG,
        get_message_info=get_message_info,
    ))


class _BasicTypesCodec:
    msg_id = BasicTypesMessage.MSG_ID

    def deserialize(self, data: bytes):
        return BasicTypesMessage.deserialize(data)


def make_sdk_with_codec(transport):
    """SDK with a BasicTypesMessage codec so handlers receive typed objects."""
    sdk = make_sdk(transport)
    sdk.register_codec(_BasicTypesCodec())
    return sdk


def encode(msg) -> bytes:
    writer = BufferWriter(PROFILE_STANDARD_CONFIG)
    writer.write(msg)
    return bytes(writer.data())


def inject_after(transport, msg, delay=0.05):
    """Inject a framed message into the transport after *delay* seconds."""
    def _work():
        time.sleep(delay)
        transport.inject_data(encode(msg))
    t = threading.Thread(target=_work, daemon=True)
    t.start()
    return t


# ---------------------------------------------------------------------------
# Test infrastructure
# ---------------------------------------------------------------------------

tests_run = 0
tests_passed = 0
tests_failed = 0


def run_test(name: str, result: bool):
    global tests_run, tests_passed, tests_failed
    tests_run += 1
    status = "PASS" if result else "FAIL"
    print(f"  {status}  {name}")
    if result:
        tests_passed += 1
    else:
        tests_failed += 1


# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

def test_basic_request_response():
    """request() returns the first response with the matching msg_id."""
    transport = MockTransport()
    sdk = make_sdk_with_codec(transport)

    req = BasicTypesMessage()
    req.regular_int = 10

    resp = BasicTypesMessage()
    resp.regular_int = 10
    resp.flag = True

    t = inject_after(transport, resp)
    result = sdk.request(req, BasicTypesMessage, timeout=2.0)
    t.join()

    run_test("basic: request() returns a result", result is not None)
    run_test("basic: returned object is the response type",
             isinstance(result, BasicTypesMessage))
    run_test("basic: response field value correct", result.flag is True)
    run_test("basic: request was sent through transport", len(transport.sent_data) == 1)


def test_timeout():
    """request() raises TimeoutError when no response arrives."""
    transport = MockTransport()
    sdk = make_sdk_with_codec(transport)

    req = BasicTypesMessage()
    req.regular_int = 99

    raised = False
    try:
        sdk.request(req, BasicTypesMessage, timeout=0.1)
    except TimeoutError:
        raised = True

    run_test("timeout: TimeoutError raised", raised)
    run_test("timeout: request was still sent", len(transport.sent_data) == 1)


def test_match_predicate_filters_responses():
    """match predicate lets caller correlate by a field value."""
    transport = MockTransport()
    sdk = make_sdk_with_codec(transport)

    req = BasicTypesMessage()
    req.regular_int = 7

    # Inject a response that does NOT match first, then one that does.
    wrong = BasicTypesMessage()
    wrong.regular_int = 99
    wrong.flag = False

    correct = BasicTypesMessage()
    correct.regular_int = 7
    correct.flag = True

    def _inject():
        time.sleep(0.03)
        transport.inject_data(encode(wrong))
        time.sleep(0.03)
        transport.inject_data(encode(correct))

    t = threading.Thread(target=_inject, daemon=True)
    t.start()
    result = sdk.request(
        req, BasicTypesMessage,
        match=lambda r: r.regular_int == 7,
        timeout=2.0,
    )
    t.join()

    run_test("match: non-matching response is ignored", result is not None)
    run_test("match: correct response returned", result.regular_int == 7)
    run_test("match: flag field is from matching response", result.flag is True)


def test_concurrent_requests_with_match():
    """Two in-flight requests with different correlation IDs resolve independently."""
    transport = MockTransport()
    sdk = make_sdk_with_codec(transport)

    results = {}
    errors = {}

    def _do_request(key, match_val, timeout=2.0):
        req = BasicTypesMessage()
        req.regular_int = match_val
        try:
            results[key] = sdk.request(
                req, BasicTypesMessage,
                match=lambda r, v=match_val: r.regular_int == v,
                timeout=timeout,
            )
        except Exception as e:
            errors[key] = e

    t1 = threading.Thread(target=_do_request, args=('a', 1), daemon=True)
    t2 = threading.Thread(target=_do_request, args=('b', 2), daemon=True)
    t1.start()
    t2.start()

    # Let both requests block, then inject responses in reverse order.
    time.sleep(0.05)
    resp_b = BasicTypesMessage(); resp_b.regular_int = 2; resp_b.flag = False
    resp_a = BasicTypesMessage(); resp_a.regular_int = 1; resp_a.flag = True
    transport.inject_data(encode(resp_b))
    transport.inject_data(encode(resp_a))

    t1.join(timeout=3.0)
    t2.join(timeout=3.0)

    run_test("concurrent: both requests resolved", len(results) == 2)
    run_test("concurrent: request 'a' got correct response",
             results.get('a') is not None and results['a'].regular_int == 1)
    run_test("concurrent: request 'b' got correct response",
             results.get('b') is not None and results['b'].regular_int == 2)
    run_test("concurrent: no errors", len(errors) == 0)


def test_request_raw_returns_bytes():
    """request_raw() returns raw payload bytes without a response class."""
    transport = MockTransport()
    sdk = make_sdk(transport)

    req = BasicTypesMessage()
    req.regular_int = 55

    resp = BasicTypesMessage()
    resp.regular_int = 55
    resp.medium_int = 123

    t = inject_after(transport, resp)
    raw = sdk.request_raw(req, BasicTypesMessage.MSG_ID, timeout=2.0)
    t.join()

    run_test("request_raw: returns bytes", isinstance(raw, (bytes, bytearray)))
    decoded = BasicTypesMessage.deserialize(raw)
    run_test("request_raw: payload round-trips correctly", decoded.medium_int == 123)


def test_request_raw_timeout():
    """request_raw() raises TimeoutError when no response arrives."""
    transport = MockTransport()
    sdk = make_sdk(transport)

    req = BasicTypesMessage()
    raised = False
    try:
        sdk.request_raw(req, BasicTypesMessage.MSG_ID, timeout=0.1)
    except TimeoutError:
        raised = True

    run_test("request_raw timeout: TimeoutError raised", raised)


def test_subscription_removed_after_success():
    """The one-shot subscription is cleaned up after request() returns."""
    transport = MockTransport()
    sdk = make_sdk_with_codec(transport)

    req = BasicTypesMessage()
    req.regular_int = 3

    resp = BasicTypesMessage()
    resp.regular_int = 3

    t = inject_after(transport, resp)
    sdk.request(req, BasicTypesMessage, timeout=2.0)
    t.join()

    # After request() the internal handler list should be empty (or only
    # contain handlers registered outside of request()).
    handlers_after = sdk.message_handlers.get(BasicTypesMessage.MSG_ID, [])
    run_test("cleanup: one-shot subscription removed after success",
             len(handlers_after) == 0)


def test_subscription_removed_after_timeout():
    """The one-shot subscription is cleaned up even after a timeout."""
    transport = MockTransport()
    sdk = make_sdk_with_codec(transport)

    req = BasicTypesMessage()
    try:
        sdk.request(req, BasicTypesMessage, timeout=0.05)
    except TimeoutError:
        pass

    handlers_after = sdk.message_handlers.get(BasicTypesMessage.MSG_ID, [])
    run_test("cleanup: one-shot subscription removed after timeout",
             len(handlers_after) == 0)


def test_with_codec():
    """If a codec is registered for the response msg_id, request() returns the decoded object."""
    transport = MockTransport()
    sdk = make_sdk(transport)

    class _Codec:
        msg_id = BasicTypesMessage.MSG_ID

        def deserialize(self, data: bytes):
            msg = BasicTypesMessage.deserialize(data)
            return {"regular_int": msg.regular_int, "flag": msg.flag}

    sdk.register_codec(_Codec())

    req = BasicTypesMessage()
    req.regular_int = 0

    resp = BasicTypesMessage()
    resp.regular_int = 0
    resp.flag = True

    t = inject_after(transport, resp)
    result = sdk.request(req, BasicTypesMessage, timeout=2.0)
    t.join()

    run_test("codec: handler receives decoded object", isinstance(result, dict))
    run_test("codec: flag field preserved", result.get("flag") is True)


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main():
    print()
    print("========================================")
    print("REQUEST/RESPONSE TESTS - Python (sync)")
    print("========================================")
    print()

    test_basic_request_response()
    test_timeout()
    test_match_predicate_filters_responses()
    test_concurrent_requests_with_match()
    test_request_raw_returns_bytes()
    test_request_raw_timeout()
    test_subscription_removed_after_success()
    test_subscription_removed_after_timeout()
    test_with_codec()

    print()
    print("========================================")
    print(f"Summary: {tests_passed}/{tests_run} tests passed")
    print("========================================")
    print()

    return 1 if tests_failed > 0 else 0


if __name__ == "__main__":
    sys.exit(main())
